### PR TITLE
get: fix halt bug + improvements

### PR DESCRIPTION
**Pull request by jbenet**
789fb77 (Juan Batiz-Benet, 5 hours ago)
    get: fix bug + improvements

    up until now there has been a very annoying bug with get, we would get
    halting behavior. I'm not 100% sure this commit fixes it, but it should. It
    certainly fixes others found in the process of digging into the get / tar
    extractor code. (wish we could repro the bug reliably enough to make a test
    case).

    This is a much cleaner tar writer. the ad-hoc, error-prone synch for the
    tar reader is gone (with i believe was incorrect). it is replaced with a
    simple pipe and bufio. The tar logic is now in tar.Writer, which writes
    unixfs dag nodes into a tar archive (no need for synch here). And get's
    reader is constructed with DagArchive which sets up the pipe + bufio.

    NOTE: this commit also changes this behavior of `get`: When retrieving a
    single file, if the file exists, get would fail. this emulated the behavior
    of wget by default, which (without opts) does not overwrite if the file is
    there. This change makes get fail if the file is available locally. This
    seems more intuitive to me as expected from a unix tool-- though perhaps it
    should be discussed more before adopting.

    Everything seems to work fine, and i have not been able to reproduce the
    get halt bug.

    License: MIT Signed-off-by: Juan Batiz-Benet <juan@benet.ai>

 c4aa506 (Juan Batiz-Benet, 5 hours ago)
    sharness/t009-get: fix space/tabs

    License: MIT Signed-off-by: Juan Batiz-Benet <juan@benet.ai>

 8e54731 (Juan Batiz-Benet, 5 hours ago)
    thirdparty/tar: clearer var names for recursion

    License: MIT Signed-off-by: Juan Batiz-Benet <juan@benet.ai>

 33a37d2 (Juan Batiz-Benet, 25 hours ago)
    unixfs/tar: cleaned up reader code

    License: MIT Signed-off-by: Juan Batiz-Benet <juan@benet.ai>

### CODE REVIEW

File: unixfs/tar/writer.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	// construct the tar writer
+	w, err := NewWriter(bufw, dag, compression)
+	if err != nil {
+		return nil, err
+	}
+
+	// write all the nodes recursively
+	go func() {
+		if err := w.WriteNode(ctx, nd, filename); err != nil {
+			pipew.CloseWithError(err)
```
Comment (whyrusleeping) [inline unixfs/tar/writer.go]: came here looking for `CloseWithError`, found it, now i'm happy.

**Comment by whyrusleeping** [comment]
kinda disappointed this didnt fix our `test_expect_failure`s, but it looks like good stuff to me!

**Comment by whyrusleeping** [comment]
> This change makes get fail if the file is available locally.

Wait, hold on. by 'available locally' do you mean, 'in our local blockstore' ?

**Comment by jbenet** [comment]
@whyrusleeping that's misleading and a bit incorrect. i meant:

> This change makes get **_no longer_** fail if **_a file is at the same path_**.

**Comment by whyrusleeping** [comment]
@jbenet ah! much better, thanks for clearing that up :)

**Pull request merged**